### PR TITLE
annotation: Add appgw-ssl-certificate annotation to allow using existing cert

### PR DESCRIPTION
**Pull request by 3quanfeng**
This PR adds a new annotation `appgw-ssl-certificate`. Using this annotation, user can reference an existing certificate in the app gateway config.
Certificates are managed at App Gateway. It could be either a base 64 pfx added directly to App Gateway or a KeyVault referenced secret (certificate).

More information will be provided in the docs.

### CODE REVIEW

File: pkg/appgw/ingress_rules.go [hunk 1]
```diff
@@ -32,18 +32,29 @@ func (c *appGwConfigBuilder) getListenersFromIngress(ingress *v1beta1.Ingress, e
...
 	listeners := make(map[listenerIdentifier]listenerAzConfig)
 
 	// Private IP is used when either annotation use-private-ip or USE_PRIVATE_IP env variable is true.
 	usePrivateIPFromAnnotation, _ := annotations.UsePrivateIP(ingress)
 	usePrivateIPForIngress := usePrivateIPFromAnnotation || env.UsePrivateIP == "true"
 
+	appgwCertName, _ := annotations.GetAppGwSslCertificate(ingress)
+	if len(appgwCertName) > 0 {
+		// logging to see the namespace of the ingress annotated with appgw-ssl-certificate
+		glog.V(5).Infof("Found anonotation appgw-ssl-certificate: %s in ingress %s/%s", appgwCertName, ingress.Namespace, ingress.Name)
```
Comment (akshaysngupta) [inline pkg/appgw/ingress_rules.go]: nit: annotation

File: pkg/appgw/ingress_rules.go [hunk 2]
```diff
@@ -53,13 +64,28 @@ func (c *appGwConfigBuilder) processIngressRule(rule *v1beta1.IngressRule, ingre
 			redirect = generateSSLRedirectConfigurationName(listenerID)
 		}
 
-		listeners[listenerID] = listenerAzConfig{
-			Protocol:                     n.HTTPS,
-			Secret:                       *secID,
-			SslRedirectConfigurationName: redirect,
+		// appgw-ssl-certificate annotation will be ignored if TLS spec found
```
Comment (akshaysngupta) [inline pkg/appgw/ingress_rules.go]: maybe:
```golang
azConfig := listenerAzConfig{
  Protocol:                     n.HTTPS,
  SslRedirectConfigurationName: redirect,
}
if cert != nil {
  azConfig.Secret: = *secID
} else if len(appgwCertName) > 0 {
  azConfig.Secret = secretIdentifier{
    Name:      appgwCertName,
    Namespace: "",
  }
}
listeners[listenerID] = azConfig
```

File: pkg/appgw/ingress_rules.go [hunk 3]
```diff
@@ -32,18 +32,29 @@ func (c *appGwConfigBuilder) getListenersFromIngress(ingress *v1beta1.Ingress, e
...
+	// certificate from ingress TLS spec
 	ingressHostnameSecretIDMap := c.newHostToSecretMap(ingress)
+
 	listeners := make(map[listenerIdentifier]listenerAzConfig)
 
 	// Private IP is used when either annotation use-private-ip or USE_PRIVATE_IP env variable is true.
 	usePrivateIPFromAnnotation, _ := annotations.UsePrivateIP(ingress)
 	usePrivateIPForIngress := usePrivateIPFromAnnotation || env.UsePrivateIP == "true"
 
+	appgwCertName, _ := annotations.GetAppGwSslCertificate(ingress)
```
Comment (akshaysngupta) [inline pkg/appgw/ingress_rules.go]: Please add a validation to check if the cert exists for ingress in https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/pkg/controller/prune.go

File: pkg/appgw/ingress_rules_test.go [hunk 4]
```diff
@@ -77,6 +86,41 @@ var _ = Describe("MutateAppGateway ingress rules, listeners, and ports", func()
 		})
 	})
 
+	Context("ingress rules with annotated certificates", func() {
```
Comment (akshaysngupta) [inline pkg/appgw/ingress_rules_test.go]: Let's also add a functional test or a unit test for config builder to check the ssl cert is correctly set in the appgw config.

File: pkg/appgw/ingress_rules.go (hunk 3 above)
Comment (3quanfeng) [reply pkg/appgw/ingress_rules.go]: > Please add a validation to check if the cert exists for ingress in https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/pkg/controller/prune.go

added a new function pruneNoSslCertificate to filter a non-match ingress

File: pkg/appgw/ingress_rules.go (hunk 2 above)
Comment (3quanfeng) [reply pkg/appgw/ingress_rules.go]: > maybe:
> azConfig := listenerAzConfig{
>   Protocol:                     n.HTTPS,
>   SslRedirectConfigurationName: redirect,
> }
> if cert != nil {
>   azConfig.Secret: = *secID
> } else if len(appgwCertName) > 0 {
>   azConfig.Secret = secretIdentifier{
>     Name:      appgwCertName,
>     Namespace: "",
>   }
> }
> listeners[listenerID] = azConfig

Updated

File: pkg/appgw/ingress_rules_test.go (hunk 4 above)
Comment (3quanfeng) [reply pkg/appgw/ingress_rules_test.go]: > Let's also add a functional test or a unit test for config builder to check the ssl cert is correctly set in the appgw config.

Added a test case to make sure http settings is configured with ssl certificate rightly when use annotation appgw-ssl-certificate

File: pkg/controller/prune.go [hunk 5]
```diff
@@ -75,14 +76,44 @@ func pruneNoPrivateIP(c *AppGwIngressController, appGw *n.ApplicationGateway, cb
...
+	var prunedIngresses []*v1beta1.Ingress
+	set := make(map[string]bool)
+	for _, installedSslCertificate := range *appGw.SslCertificates {
+		set[*installedSslCertificate.Name] = true
+	}
+
+	for _, ingress := range ingressList {
+		annotatedSslCertificate, err := annotations.GetAppGwSslCertificate(ingress)
+		if err != nil && annotations.IsInvalidContent(err) {
+			glog.Errorf("Ingress %s/%s has invalid value for annotation %s", ingress.Namespace, ingress.Name, annotations.AppGwSslCertificate)
```
Comment (akshaysngupta) [inline pkg/controller/prune.go]: I think we can improve this error message.

File: pkg/controller/prune.go [hunk 6]
```diff
@@ -75,14 +76,44 @@ func pruneNoPrivateIP(c *AppGwIngressController, appGw *n.ApplicationGateway, cb
...
+		}
+
+		if _, exists := set[annotatedSslCertificate]; !exists {
+			errorLine := fmt.Sprintf("ignoring Ingress %s/%s as it requires Application Gateway %s to have pre-installed ssl certificate", ingress.Namespace, ingress.Name, c.appGwIdentifier.AppGwName)
```
Comment (akshaysngupta) [inline pkg/controller/prune.go]: Maybe change this to:
`
ignoring Ingress %s/%s as ssl-certificate '%s' specified by the <replace with annotation key> was not found on Application Gateway '%s'
`

File: pkg/controller/prune.go [hunk 7]
```diff
@@ -75,14 +76,44 @@ func pruneNoPrivateIP(c *AppGwIngressController, appGw *n.ApplicationGateway, cb
...
 func pruneRedirectWithNoTLS(c *AppGwIngressController, appGw *n.ApplicationGateway, cbCtx *appgw.ConfigBuilderContext, ingressList []*v1beta1.Ingress) []*v1beta1.Ingress {
 	var prunedIngresses []*v1beta1.Ingress
 	for _, ingress := range ingressList {
-		hasTLS := ingress.Spec.TLS != nil && len(ingress.Spec.TLS) > 0
+		appgwCertName, _ := annotations.GetAppGwSslCertificate(ingress)
+		hasTLS := (ingress.Spec.TLS != nil && len(ingress.Spec.TLS) > 0) || len(appgwCertName) > 0
 		sslRedirect, _ := annotations.IsSslRedirect(ingress)
 		if !hasTLS && sslRedirect {
-			errorLine := fmt.Sprintf("ignoring Ingress %s/%s as it has an invalid spec. It is annotated with ssl-redirect: true but is missing a TLS secret. Please add a TLS secret or remove ssl-redirect annotation", ingress.Namespace, ingress.Name)
+			errorLine := fmt.Sprintf("ignoring Ingress %s/%s as it has an invalid spec. It is annotated with ssl-redirect: true but is missing a TLS secret or annotation. Please add a TLS secret/annotation or remove ssl-redirect annotation", ingress.Namespace, ingress.Name)
```
Comment (akshaysngupta) [inline pkg/controller/prune.go]: Change this to TLS secret or appgw-ssl-name annotation.

File: pkg/controller/prune_test.go [hunk 8]
```diff
@@ -71,6 +71,42 @@ var _ = Describe("prune function tests", func() {
...
+			Expect(len(prunedIngresses)).To(Equal(0))
+		})
+
+		It("keeps the ingress using appgw-ssl-certificate when annotated ssl cert is pre-installed", func() {
+			ingressSslCertAnnotated.Annotations = map[string]string{
+				annotations.AppGwSslCertificate: *fixtures.GetCertificate1().Name,
+			}
+
+			Expect(len(cbCtx.IngressList)).To(Equal(2))
+			prunedIngresses := pruneNoPrivateIP(controller, &appGw, cbCtx, cbCtx.IngressList)
```
Comment (akshaysngupta) [inline pkg/controller/prune_test.go]: `pruneNoSslCertificate`

**Pull request merged**